### PR TITLE
[orc8r][service_registry] Implement service registry service for orc8r service mesh

### PR DESCRIPTION
**Pull request by mpgermano**
Signed-off-by: mgermano <mgermano@fb.com>

## Summary

This PR implements the service registry service that will support service discovery
for the orc8r service mesh. This registry has two modes, set via env-var to back the registry
by either Docker or Kubernetes. This will allow us to continue to develop with docker-compose.

A stacked PR will add the necessary changes to the service registry client to flip this service to
in-use. This will also include the necessary helm chart changes that are needed.

## Test Plan

Added unit test.

For docker, add labels to `controller` container. Use a CLI to ensure
that each RPC call returns the expected values.

For K8s, deploy the code to minikube, create necessary service-reader role, serviceAccount, and roleBinding.
Use CLI to ensure each RPC call returns the expected values.

### CODE REVIEW

File: orc8r/cloud/go/services/service_registry/servicers/docker_servicer.go [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+		services = append(services, serviceName)
+	}
+	return &protos.ListAllServicesResponse{
+		Services: services,
+	}, nil
+}
+
+// FindServices returns all services in that have the provided label.
+func (s *DockerServiceRegistryServicer) FindServices(ctx context.Context, req *protos.FindServicesRequest) (*protos.FindServicesResponse, error) {
+	if req == nil {
```
Comment (uri200) [inline orc8r/cloud/go/services/service_registry/servicers/docker_servicer.go:76]: Not sure if this is an implementation of GRPC service. But if so, is it possible req is ever nil, and we may not need this? Will not the GRPC client always create a default req?

If not GRPC implementation, then disregard the comment

File: orc8r/cloud/docker/docker-compose.yml [hunk 2]
```diff
@@ -34,7 +34,7 @@ services:
       SQL_DRIVER: postgres
       DATABASE_SOURCE: "dbname=magma_dev user=magma_dev password=magma_dev host=postgres sslmode=disable"
       SQL_DIALECT: psql
-
+      SERVICE_REGISTRY_MODE: "0" # Docker backed service registry
```
Comment (xjtian) [inline orc8r/cloud/docker/docker-compose.yml]: let's use string values for this: `docker`, `k8s`, `yml`, etc

File: orc8r/cloud/go/services/service_registry/service_registry/main.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+func main() {
+	srv, err := service.NewOrchestratorService(orc8r.ModuleName, service_registry.ServiceName)
+	if err != nil {
+		glog.Fatalf("Error creating service registry service %s", err)
+	}
+	registryModeEnvValue := os.Getenv(service_registry.ServiceRegistryModeEnvVar)
+
+	var serviceRegistryServicer protos.ServiceRegistryServer
+	switch registryModeEnvValue {
+	case "0":
```
Comment (xjtian) [inline orc8r/cloud/go/services/service_registry/service_registry/main.go]: let's make these strings and declare them as consts somewhere

File: orc8r/cloud/go/services/service_registry/servicers/docker_servicer.go [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+		return &protos.ListAllServicesResponse{}, err
+	}
+	services := []string{}
+	for _, container := range allContainers {
+		if len(container.Names) == 0 {
+			continue
+		}
+		serviceName := container.Names[0]
+		serviceName = strings.TrimPrefix(serviceName, "/")
+		services = append(services, serviceName)
```
Comment (xjtian) [inline orc8r/cloud/go/services/service_registry/servicers/docker_servicer.go]: some context here as to what the docker API returns and why you're doing the postprocessing would be useful

File: orc8r/cloud/go/services/service_registry/servicers/docker_servicer.go [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+		return &protos.FindServicesResponse{}, err
+	}
+	services := []string{}
+	for _, container := range allContainers {
+		if len(container.Names) == 0 {
+			continue
+		}
+		serviceName := container.Names[0]
+		serviceName = strings.TrimPrefix(serviceName, "/")
+		services = append(services, serviceName)
```
Comment (xjtian) [inline orc8r/cloud/go/services/service_registry/servicers/docker_servicer.go]: pull this code out into a helper, it's repeated from above

File: orc8r/cloud/go/services/service_registry/servicers/kubernetes_servicer.go [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+		LabelSelector: fmt.Sprintf("%s=%s", partOfLabel, "orc8r"),
+	}
+	svcList, err := s.client.Services(s.namespace).List(orc8rListOption)
+	if err != nil {
+		return ret, err
+	}
+	for _, svc := range svcList.Items {
+		splitSvcName := strings.SplitAfterN(svc.Name, serviceNameDelimiter, 2)
+		svcNameIndex := len(splitSvcName) - 1
+		ret.Services = append(ret.Services, splitSvcName[svcNameIndex])
```
Comment (xjtian) [inline orc8r/cloud/go/services/service_registry/servicers/kubernetes_servicer.go]: same as for the other servicer, factor this out and add some context as to what you're doing here

**Review by xjtian — CHANGES_REQUESTED** [review]

**Review by xjtian — APPROVED** [review]

**Pull request merged**